### ISSUE TITLE
Compile error with -Wconversion

### ISSUE THREAD

**Issue by wr786** [ktprime/emhash#36]
both `emhash7::HashMap` and `emilib2::HashMap` have this problem

**Comment by ktprime** [comment]
what's kind of compiler used?  try use c++17 to compile and run the bench code.

**Comment by wr786** [comment]
> what's kind of compiler used? try use c++17 to compile and run the bench code.

g++ 9.4.0, I can successfully make all bench code without errors, but

https://github.com/ktprime/emhash/blob/0820f65006dc0b98bdf922a4393297ec5c68bcb3/bench/Makefile#L10

try add `-Wconversion` after `-march=native` then `make all`, then you could see lots of noisy errors.

**Comment by ktprime** [comment]
It's seem no any errors in my evn.    

#make BF=1 SW=1 CXX=g++-9

g++-9 -O3 -DNDEBUG=1 -march=native -Wconversion -I.. -I../thirdparty -DABSL_HMAP=1 -DHAVE_BOOST=1 -std=c++17 -flto  martin_bench.cpp -o mbench && ./mbench 

-----------------------------------------------------------------------------------------------------------------
gcc 9.5.0 __cplusplus = 201703 x86-64 OS = linux, cpu = 12th Gen Intel(R) Core(TM) i7-12700
-----------------------------------------------------------------------------------------------------------------
usage: ./mbench [2-9mptseabrjqf]b[d]e[d]
all test case:
     1 bench_IterateIntegers
     2 bench_randomFindString
     3 bench_randomEraseString
     4 bench_copy
     5 bench_randomFind
     6 bench_insert
     7 bench_randomInsertErase
     8 bench_randomDistinct2
     9 bench_knucleotide
    10 bench_GameOfLife
    11 bench_AccidentallyQuadratic
    12 bench_InsertEraseContinue
    13 bench_InsertEraseBegin
    14 bench_CreateInsert
    15 bench_Udb3
-------------------------------------------------------------------------
test with max_load_factor = 0.80
all test hashmap:
      absl            absl flat
    ankerl         martin dense
     boost           boost flat
   emhash5              emhash5
   emhash6              emhash6
   emhash7              emhash7
   emhash8              emhash8
    emilib              emilib1
   emilib2              emilib2
   emilib3              emilib3
-------------------------------------------------------------------------

bench_IterateIntegers:
                 emhash5, add/removing time = 5.48, 7.38|62498750000000
                 emhash8, add/removing time = 0.30, 0.27|62498750000000
                 emhash7, add/removing time = 0.88, 0.97|62498750000000
                 emhash6, add/removing time = 0.94, 0.97|62498750000000
            martin dense, add/removing time = 0.25, 0.25|62498750000000
                 emilib1, add/removing time = 1.48, 1.56|62498750000000
                 emilib3, add/removing time = 1.33, 1.58|62498750000000
                 emilib2, add/removing time = 1.73, 1.72|62498750000000
              boost flat, add/removing time = 1.88, 2.52|62498750000000
               absl flat, add/removing time = 3.36, 4.27|62498750000000

**Comment by ktprime** [comment]
"-Werror" is too stritct and many third-party libs will be failed.

**Comment by wr786** [comment]
Yes, but I hope at least emhash can pass this, it would be safer

or use a workaround if you don't want to change, by adding the code below
```cpp
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wconversion"

/*
* Acutal Code
*/

#pragma GCC diagnostic pop
```

This would be better for those who want to introduce emhash into their projects (with `-Werror -Wconversion etc.`)

**Comment by ktprime** [comment]
I need some time to fix all warnings in my emhash code.
 btw do you use it in your project ?

**Comment by wr786** [comment]
yes, there are lots of `-W...` in my project,
and when I introduce emhash for faster speed (and indeed it's much faster), the compiler told me that `-Wconversion` failed 😢 
now I'm using the workaround above to prevent compile errors

**Comment by ktprime** [comment]
I have made some fixes to address warnings, although I'm not certain if all of them have been resolved. You can try using one of my other hash maps, such as [emilib2o](https://github.com/ktprime/emhash/blob/master/thirdparty/emilib/emilib2o.hpp) or emilib2s. These hash maps utilize SIMD instructions similar to Google's SwissTable, and they provide further optimizations for insertion and erasure operations.

**Comment by wr786** [comment]
Oh, thank you! 
By the way, what's the difference between `emilib2o` and `emilib2s`? 🤔

**Comment by ktprime** [comment]
emilib2o foucus on finding and insertion-ersion use case, a lillte more memory than emilib2s(fast on insertion)

**Comment by wr786** [comment]
![Image](https://github.com/user-attachments/assets/692c17e7-44c1-4eac-b994-ecb609da87b8)

Thank you for your efforts so far!

Just a quick note, the `-Wconversion` warnings in `emhash7` are still present.

**Comment by ktprime** [comment]
line 148 you can change it to ` #define EMH_SET(n)        _bitmask[n / MASK_BIT] &= (uint8_t)(~(1 << (n % MASK_BIT)))`
if no any errors, I'll commit again. I can not reproduce it and what's complier you used and full compiler options?

**Comment by wr786** [comment]
one way to reproduce this problem is using [GitHub Codespace](https://github.com/ktprime/emhash/codespaces), which provides an online environment that ensures consistency across all users.

![Image](https://github.com/user-attachments/assets/208f45fb-8995-4d73-b02a-81bdb03b22c6)

I’ve tested and confirmed that the problem can be reproduced in the Codespace.

![Image](https://github.com/user-attachments/assets/fd9f0737-f7bd-4d66-84dc-4def9de03dcc)

Unfortunately, while the issue with this line has been resolved, it seems to have revealed additional `-Wconversion` warnings.

**Comment by ktprime** [comment]
Try again, the above error is test code issue.

**Comment by wr786** [comment]
I just re-ran the build, and I’m delighted to announce that all the `-Wconversion` warnings in emhash have been resolved! 🎉
